components/search/src/filter.rs

```diff
@@ -108,11 +110,87 @@ fn maybe_extract_engine_config(
         .map(|_variant| SearchEngineDefinition::from_configuration_details(&identifier, base))
 }
 
+fn determine_default_engines(
+    engines: &[SearchEngineDefinition],
+    default_engines_record: Option<JSONDefaultEnginesRecord>,
+    user_environment: &SearchUserEnvironment,
+) -> (Option<String>, Option<String>) {
+    match default_engines_record {
+        None => (None, None),
+        Some(record) => {
+            let mut default_engine_id = None;
+            let mut default_engine_private_id = None;
+
+            for data in record.specific_defaults {
+                if matches_user_environment(&data.environment, user_environment) {
+                    // Check the engine is present in the list of engines before
+                    // we return it as default.
+                    let engine_id = find_engine_with_match(engines, data.default);
+                    if engine_id.is_some() {
+                        default_engine_id = engine_id;
+                    }
+                    let private_engine_id = find_engine_with_match(engines, data.default_private);
+                    if private_engine_id.is_some() {
+                        default_engine_private_id = private_engine_id;
+                    }
+                }
+            }
```

Comment:
```suggestion
            let (default_engine_id, default_engine_private_id) = record.specific_defaults
                .into_iter()
                .filter(|data| matches_user_environment(&data.environment, user_environment))
                .fold((None, None), |(default_engine_id, default_engine_private_id), data| {
                    (
                        // Check the engine is present in the list of engines before
                        // we return it as default.
                        find_engine_with_match(engines, data.default).or(default_engine_id),
                        find_engine_with_match(engines, data.default_private).or(default_engine_private_id),
                    )
                });
```

I like using `fold()` to extract multiple bits of data like this, but either way is great with me—whichever you like more!